### PR TITLE
Sphinx doc chapter 15

### PR DESCRIPTION
**Pull request by maximedenes**

### CODE REVIEW

File: doc/sphinx/practical-tools/utilities.rst [hunk 1]
```diff
@@ -1,482 +1,981 @@
...
-Slow (user: 0.36 mem: 393912 ko)
-COQC Fast.v
-Fast (user: 0.05 mem: 45992 ko)
-Time     | File Name
---------------------
-0m00.41s | Total
+
+  ::
+
+    pre-all::
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: These two colons comes from the original text. Unfortunately they seem to be interpreted as starting some verbatim text and should probably be quoted. The same is true for the following targets. Note that I didn't test this but I'm basing my assumption on GitHub's syntax highlighting.

File: doc/sphinx/practical-tools/utilities.rst [hunk 2]
```diff
@@ -1,482 +1,981 @@
...
+++++++++++++++++++++++++++++++++++++++
+
+The generated ``Makefile`` supports the generation of two kinds of timing
+data: per-file build-times, and per-line times for an individual file.
+
+The following targets and Makefile variables allow collection of per-
+file timing data:
+
+
++ TIMED=1
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim formatting was lost.

File: doc/sphinx/practical-tools/utilities.rst [hunk 3]
```diff
@@ -1,482 +1,981 @@
...
+
+      COQDEP Fast.v
+      COQDEP Slow.v
+      COQC Slow.v
+      Slow (user: 0.34 mem: 395448 ko)
+      COQC Fast.v
+      Fast (user: 0.01 mem: 45184 ko)
+
+
++ pretty-timed
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim formatting was lost.

File: doc/sphinx/practical-tools/utilities.rst [hunk 4]
```diff
@@ -1,482 +1,981 @@
...
+    this target stores the output of ``make TIMED=1`` into
+    ``time-of-build.log``, and displays a table of the times, sorted from
+    slowest to fastest, which is also stored in ``time-of-build-pretty.log``.
+    If you want to construct the ``log`` for targets other than the default
+    one, you can pass them via the variable ``TGTS``, e.g., ``make pretty-timed
+    TGTS="a.vo b.vo"``. 
+
+    Note: This target requires ``python`` to build the
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Should probably use the standard `..remark::`

File: doc/sphinx/practical-tools/utilities.rst [hunk 5]
```diff
@@ -1,482 +1,981 @@
...
+    table. 
+
+    Note: This target will *append* to the timing log; if you want a
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Same.

File: doc/sphinx/practical-tools/utilities.rst [hunk 6]
```diff
@@ -1,482 +1,981 @@
...
+    fresh start, you must remove the ``filetime-of-build.log`` or ``run make
+    cleanall``. 
+
+    Example: For example, the output of ``make pretty-timed`` may
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Should probably use the standard `..example::`

File: doc/sphinx/practical-tools/utilities.rst [hunk 7]
```diff
@@ -1,482 +1,981 @@
...
+      Fast (user: 0.05 mem: 45992 ko)
+      Time     | File Name
+      --------------------
+      0m00.41s | Total
+      --------------------
+      0m00.36s | Slow
+      0m00.05s | Fast
+
+
++ print-pretty-timed-diff
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim formatting was lost.

File: doc/sphinx/practical-tools/utilities.rst [hunk 8]
```diff
@@ -1,482 +1,981 @@
...
+    this target builds a table of timing
+    changes between two compilations; run ``make make-pretty-timed-before`` to
+    build the log of the “before” times, and run ``make make-pretty-timed-
+    after`` to build the log of the “after” times. The table is printed on
+    the command line, and stored in ``time-of-build-both.log``. This target is
+    most useful for profiling the difference between two commits to a
+    repo.
+
+    Note: This target requires ``python`` to build the table.
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Same remark on many following "Note" and "Example"

File: doc/sphinx/practical-tools/utilities.rst [hunk 9]
```diff
@@ -1,482 +1,981 @@
...
+      0m00.37s | Slow      | 0m00.01s || +0m00.36s | +3600.00%
+      0m00.02s | Fast      | 0m00.34s || -0m00.32s | -94.11%
+
+
+
+The following targets and ``Makefile`` variables allow collection of per-
+line timing data:
+
+
++ TIMING=1
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim

File: doc/sphinx/practical-tools/utilities.rst [hunk 10]
```diff
@@ -1,482 +1,981 @@
...
+    Example: For example, running ``make all TIMING=1`` may result in a file like this:
+
+    ::
+
+        Chars 0 - 26 [Require~Coq.ZArith.BinInt.] 0.157 secs (0.128u,0.028s)
+        Chars 27 - 68 [Declare~Reduction~comp~:=~vm_c...] 0. secs (0.u,0.s)
+        Chars 69 - 162 [Definition~foo0~:=~Eval~comp~i...] 0.153 secs (0.136u,0.019s)
+        Chars 163 - 208 [Definition~foo1~:=~Eval~comp~i...] 0.239 secs (0.236u,0.s)
+
++ print-pretty-single-time-diff BEFORE=path/to/file.v.before-timing AFTER=path/to/file.v.after-timing
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim

File: doc/sphinx/practical-tools/utilities.rst [hunk 11]
```diff
@@ -1,482 +1,981 @@
...
+      0m00.50s  | Total                                               | 0m04.17s  || -0m03.66s | -87.96%
+      ---------------------------------------------------------------------------------------------------
+      0m00.145s | Chars 069 - 162 [Definition~foo0~:=~Eval~comp~i...] | 0m00.192s || -0m00.04s | -24.47%
+      0m00.126s | Chars 000 - 026 [Require~Coq.ZArith.BinInt.]        | 0m00.143s || -0m00.01s | -11.88%
+         N/A    | Chars 027 - 068 [Declare~Reduction~comp~:=~nati...] | 0m00.s    || +0m00.00s | N/A
+      0m00.s    | Chars 027 - 068 [Declare~Reduction~comp~:=~vm_c...] |    N/A    || +0m00.00s | N/A
+      0m00.231s | Chars 163 - 208 [Definition~foo1~:=~Eval~comp~i...] | 0m03.836s || -0m03.60s | -93.97%
+
+
++ all.timing.diff, path/to/file.v.timing.diff
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Verbatim

File: doc/sphinx/practical-tools/utilities.rst [hunk 12]
```diff
@@ -1,482 +1,981 @@
...
+
+See the man page of ``coqdep`` for more details and options.
+
+The build infrastructure generated by ``coq_makefile`` uses ``coqdep`` to
+automatically compute the dependencies among the files part of the
+project.
+
+
+Documenting |Coq| files with coqdoc
+-----------------------------------
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: This whole section comes from another file which was not removed in this PR.

File: doc/sphinx/practical-tools/utilities.rst [hunk 13]
```diff
@@ -1,482 +1,981 @@
...
-  \ttindex{coq-tex}\index{Latex@{\LaTeX}}}
+
+
+Sections.
++++++++++
+
+Sections are introduced by 1 to 4 leading stars (i.e. at the beginning
+of the line) followed by a space. One star is a section, two stars a
+sub-section, etc. The section title is given on the remaining of the
+line. Example:
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: `..example::`

File: doc/sphinx/practical-tools/utilities.rst [hunk 14]
```diff
@@ -1,482 +1,981 @@
...
+
+List items are introduced by a leading dash. coqdoc uses whitespace to
+determine the depth of a new list item and which text belongs in which
+list items. A list ends when a line of text starts at or before the
+level of indenting of the list’s dash. A list item’s dash must always
+be the first non-space character on its line (so, in particular, a
+list can not begin on the first line of a comment - start it on the
+second line instead).
+
+Example:
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Same

File: doc/sphinx/practical-tools/utilities.rst [hunk 15]
```diff
@@ -1,482 +1,981 @@
...
+
++ ``$...LATEX stuff...$`` inserts some |Latex| material in math mode.
+  Simply discarded in |HTML| output.
++ ``%...LATEX stuff...%`` inserts some |Latex| material. Simply
+  discarded in |HTML| output.
++ ``#...HTML stuff...#`` inserts some |HTML| material. Simply discarded in
+  |Latex| output.
+
+
+Note: to simply output the characters ``$``, ``%`` and ``#`` and escaping
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: `..remark::`

**Comment by maximedenes** [comment]
@thery do you think you could address the comments of @Zimmi48 ?

File: doc/sphinx/practical-tools/utilities.rst (hunk 12 above)
Comment (thery) [reply doc/sphinx/practical-tools/utilities.rst]: @Zimmi48 I don't understand what I should do

**Comment by thery** [comment]
@maximedenes I must have been tired when I did that chapter. I didn't do a good job...
AnywayI did a pull request to maximedenes/coq with all the corrections by @Zimmi48.
The only one I didn't take into account  is the comment about the section 
`Documenting |Coq| files with coqdoc`. 
I didn't understand it

**Comment by Zimmi48** [comment]
I'm just saying that this whole part about coqdoc comes from another file `doc/refman/coqdoc.tex` and this PR should remove it since it is migrated to Sphinx. But @maximedenes can certainly take care of it when updating this PR following your corrections.

File: doc/sphinx/practical-tools/utilities.rst [hunk 16]
```diff
@@ -1,482 +1,1054 @@
...
+For example, to statically link |L_tac|, you can add it to ``tools/tolink.ml`` and use:
+
+::
+
+   % make tools/tolink.cmx
+   % ocamlfind ocamlopt -thread -rectypes -linkall -linkpkg \
+                 -package coq.toplevel -package coq.ltac    \
+                 -I tools tolink.cmx toplevel/coqtop\_bin.ml -o my\_toplevel.native
+
+We will remove the need for the ``tolink`` file in the future.
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Sorry but it looks once again like what was ported wasn't the most recent version of this section. See the diff.

File: doc/sphinx/practical-tools/utilities.rst [hunk 17]
```diff
@@ -1,482 +1,981 @@
...
-The {\tt coq\_makefile} utility can be used to set up a build infrastructure
-for the \Coq\ project based on makefiles.  The recommended way of
-invoking {\tt coq\_makefile} is the following one:
-
-\begin{verbatim}
-coq_makefile -f _CoqProject -o CoqMakefile
-\end{verbatim}
+
+Building a toplevel extended with user tactics
+----------------------------------------------
```
[the remark this answers is not in the document]
Comment (Zimmi48) [reply doc/sphinx/practical-tools/utilities.rst]: This section still needs to be removed.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]

File: doc/sphinx/practical-tools/utilities.rst [hunk 18]
```diff
@@ -0,0 +1,1054 @@
...
+
+   % make tools/tolink.cmx
+   % ocamlfind ocamlopt -thread -rectypes -linkall -linkpkg \
+                 -package coq.toplevel -package coq.ltac    \
+                 -I tools tolink.cmx toplevel/coqtop\_bin.ml -o my\_toplevel.native
+
+We will remove the need for the ``tolink`` file in the future.
+
+Building a toplevel extended with user tactics
+----------------------------------------------
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: This section still needs to be removed.

**Comment by maximedenes** [comment]
@thery Can you address the last comments by @Zimmi48 ? Thanks!

**Comment by thery** [comment]
@Zimmi48 colpa mia
@maximedenes  done

File: doc/sphinx/practical-tools/utilities.rst [hunk 19]
```diff
@@ -1,482 +1,1032 @@
...
-for the \Coq\ project based on makefiles.  The recommended way of
-invoking {\tt coq\_makefile} is the following one:
-
-\begin{verbatim}
-coq_makefile -f _CoqProject -o CoqMakefile
-\end{verbatim}
+
+
+Application: how to use the |OCaml| debugger with |Coq|.
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst]: Sorry, but this part still needs to be removed.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
Remove the remaining outdated part, do `git rm doc/refman/coqdoc.tex && git commit --amend --no-edit`, rebase and fix the trivial conflicts in the Makefile and we should be all good.

**Comment by thery** [comment]
@Zimmi48 hope I haven't been stuttering again....

File: Makefile.doc [hunk 20]
```diff
@@ -65,7 +65,7 @@ REFMANCOQTEXFILES:=$(addprefix doc/refman/, \
 
 REFMANTEXFILES:=$(addprefix doc/refman/, \
   headers.sty Reference-Manual.tex \
-  AsyncProofs.tex) \
+  RefMan-uti.tex) \
```
Comment (Zimmi48) [inline Makefile.doc]: There is still an issue here (must have happened during the rebase).
This file shouldn't have changed in this commit.

Comment (thery) [reply Makefile.doc]: @Zimmi48 revert the change of the makefile

Comment (Zimmi48) [reply Makefile.doc]: What do you mean exactly?

**Comment by thery** [comment]
@Zimmi48 Ok I remove my change of the Makefile

**Comment by Zimmi48** [comment]
You're still adding a newline which basically amounts to reverting the change to the Makefile done in the previous commit.
Sorry to bother you so much!

**Comment by thery** [comment]
@Zimmi48 I don't see the change of Makefile in the pull request so I've assumed it was ok

**Comment by Zimmi48** [comment]
That's precisely the problem. The first commit removes something from the Makefile and this is wanted. The second commit reverts this change and this is unwanted.

**Review by Zimmi48 — APPROVED** [review]
I'll merge this as soon as CI is green.

**Comment by thery** [comment]
@Zimmi48 thanks!

**Pull request merged**